src/mongocrypt-marking.c

```diff
@@ -1164,72 +1162,130 @@ static bool _fle2_generate_TextSearchTokenSets(_mongocrypt_key_broker_t *kb,
         goto fail;
     }
 
-    if (!_fle2_generate_TextExactTokenSet(kb,
-                                          &tsts->exact,
-                                          &foldedValue,
-                                          contentionFactor,
-                                          common.collectionsLevel1Token,
-                                          common.serverTokenDerivationLevel1Token,
-                                          status)) {
-        goto fail;
+    // Generate exact token set singleton
+    {
+        _mongocrypt_buffer_t asBsonValue;
+        _mongocrypt_buffer_init(&asBsonValue);
+        BSON_ASSERT(encodeSets->exact.len < INT_MAX);
+        _mongocrypt_buffer_copy_from_string_as_bson_value(&asBsonValue,
+                                                          (const char *)encodeSets->exact.data,
+                                                          (int)encodeSets->exact.len);
+        if (!_fle2_generate_TextExactTokenSet(kb,
+                                              &tsts->exact,
+                                              &asBsonValue,
+                                              contentionFactor,
+                                              common.collectionsLevel1Token,
+                                              common.serverTokenDerivationLevel1Token,
+                                              status)) {
+            goto fail;
+        }
+        _mongocrypt_buffer_cleanup(&asBsonValue);
     }
 
+    const char *substring;
+    uint32_t bytelen;
+    uint32_t appendCount;
+
+    // Generate array of substring token sets
     if (spec->substr.set) {
-        // TODO MONGOCRYPT-762 iterate on StrEncode substrings set
-        mc_TextSubstringTokenSet_t substrSet = {{0}};
-        mc_TextSubstringTokenSet_init(&substrSet);
-
-        if (!_fle2_generate_TextSubstringTokenSet(kb,
-                                                  &substrSet,
-                                                  &foldedValue,
-                                                  contentionFactor,
-                                                  common.collectionsLevel1Token,
-                                                  common.serverTokenDerivationLevel1Token,
-                                                  status)) {
-            mc_TextSubstringTokenSet_cleanup(&substrSet);
-            goto fail;
+        mc_substring_set_iter_t set_itr;
+        mc_substring_set_iter_init(&set_itr, encodeSets->substring_set);
+
+        while (mc_substring_set_iter_next(&set_itr, &substring, &bytelen, &appendCount)) {
+            for (; appendCount > 0; appendCount--) {
```

Comment:
Let's move everything from 1196-1214 outside the loop, and then append tset multiple times. I don't think it's a problem that the copies all point at the same underlying buffer, since from this point the token sets should be read-only.

Reply:
Done, but the array append just does a memcpy of the `mc_Text<?>TokenSet_t`, so we have to make a deep copy of it and the mongocrypt_buffers within it otherwise we end up with multiple owning mongocrypt_buffers pointing to the same heap address.